### PR TITLE
Implement partial & repeat ties

### PR DESCRIPTION
**Pull request by miiizen**
Resolves: https://github.com/musescore/MuseScore/issues/25426

This PR introduces partial ties to MuseScore, enabling ties to be connected across repeats. When adding a tie before a repeat item (repeat barlines, markers, jumps, and voltas) if any note on the other side of the repeat destination can be tied, a partial tie will be added.  
When more than one endpoint is available, clicking the outgoing tie will open a menu of possible endpoints.  These can be toggled on and off.  

Screenshot 2024-12-05 at 13 06 28

Screenshot 2024-12-05 at 13 07 54

https://github.com/user-attachments/assets/6c1c5120-350b-418c-8052-a9e4f5621e82

Playback implementation to follow.

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/compat/midi/compatmidirender.cpp [hunk 1]
```diff
@@ -1080,7 +1080,7 @@ std::set<size_t> CompatMidiRender::getNotesIndexesToRender(Chord* chord)
     }
 
     auto noteShouldBeRendered = [](Note* n) {
-        while (n->tieBack() && n != n->tieBack()->startNote()) {
+        while (n->tieBack() && !n->incomingPartialTie() && n != n->tieBack()->startNote()) {
```
Comment (mike-spa) [inline src/engraving/compat/midi/compatmidirender.cpp:1083]: I've seen this check repeated several times: every time you are looking for a tie-back-but-not-incoming-partial. Perhaps this is worth adding a `tieBackNonPartial()` method?

File: src/engraving/dom/tie.cpp [hunk 2]
```diff
@@ -213,6 +212,148 @@ Tie::Tie(const ElementType& type, EngravingItem* parent)
...
+    }
+    for (Tie* tie : oldTies) {
+        auto findEndTie = [&tie](const TieEndPoint* endPoint){
+            return endPoint->endTie() == tie;
+        };
+
+        if (std::find_if((*tieEndPoints()).begin(), (*tieEndPoints()).end(), findEndTie) != (*tieEndPoints()).end()) {
+            continue;
+        }
+        score()->undoRemoveElement(tie);
```
Comment (mike-spa) [inline src/engraving/dom/tie.cpp:288]: A method of the Tie class should not be removing other ties from the Score, it could be a bit dangerous. Also, from a function named collectPossibleEndPoints I'd probably expect something which returns me a list of TieEndPoint*, not something which actively modifies the current tie and even the score. Here I would try to do a combination of (not necessarily all of them)

- Giving this method a more descriptive name
- Moving out the code which removes other ties into a separate function
- Given that this method performs an undoable action on the score, it looks more like the kind of methods we usually put in edit.cpp, that's also an option

File: src/engraving/dom/tie.h [hunk 3]
```diff
@@ -25,6 +25,64 @@
 #include "slurtie.h"
 
 namespace mu::engraving {
+class TieEndPointList;
+class TieEndPoint
```
Comment (mike-spa) [inline src/engraving/dom/tie.h:29]: Side question: would TieJumpPoint be perhaps a better name for this class? I know what you mean by TieEndPoint, but it feels to me like it doesn't communicate the main purpose. Also because we often call "end point" the graphical end points too.

File: src/engraving/dom/note.cpp [hunk 4]
```diff
@@ -1282,15 +1304,25 @@ void Note::add(EngravingItem* e)
...
         lv->setStartNote(this);
         lv->setTick(lv->startNote()->tick());
-        lv->setTrack(track());
         setTieFor(lv);
         break;
     }
+    case ElementType::PARTIAL_TIE: {
+        PartialTie* pt = toPartialTie(e);
+        pt->setTick(tick());
+        if (pt->partialSpannerDirection() == PartialSpannerDirection::OUTGOING) {
```
Comment (mike-spa) [inline src/engraving/dom/note.cpp:1313]: `pt->itOutGoing()` ?

File: src/engraving/dom/excerpt.cpp [hunk 5]
```diff
@@ -770,15 +771,15 @@ static void processLinkedClone(EngravingItem* ne, Score* score, track_idx_t stra
...
+    if (originalNote->tieFor() && originalNote->tieFor()->type() == ElementType::TIE) {
         Tie* tie = toTie(originalNote->tieFor()->linkedClone());
         tie->setScore(score);
         newNote->setTieFor(tie);
         tie->setStartNote(newNote);
         tie->setTrack(newNote->track());
         tieMap.add(originalNote->tieFor(), tie);
     }
-    if (originalNote->tieBack()) {
+    if (originalNote->tieBack() && originalNote->tieBack()->type() == ElementType::TIE) {
```
[the remark this answers is not in the document]
Comment (miiizen) [reply src/engraving/dom/excerpt.cpp:782]: I considered this, but it seemed to me that there were far more scenarios where we want `tieFor` and `tieBack` to return LVs & partial ties than when we didn't.

File: src/engraving/dom/tie.cpp (hunk 2 above)
Comment (miiizen) [reply src/engraving/dom/tie.cpp:288]: Yes, this code to remove ties is absolutely in the wrong place - it doesn't even cover all the situations I was hoping.  
I think we need to remove stale endpoints when the repeat structure is recalculated - either inside or just after `RepeatList::unwind`.

File: src/framework/uicomponents/qml/Muse/UiComponents/internal/StyledMenu.qml [hunk 6]
```diff
@@ -32,6 +32,9 @@ MenuView {
 
     property alias model: view.model
 
+    property NavigationSection notationViewNavigationSection: null
```
Comment (cbjeukendrup) [inline src/framework/uicomponents/qml/Muse/UiComponents/internal/StyledMenu.qml:35]: The `UiComponents` module should not contain the word "notation", because it is intended to be as general as possible. Something like `navigationSectionOverride` would be okay though.

File: src/framework/uicomponents/qml/Muse/UiComponents/internal/StyledMenu.qml [hunk 7]
```diff
@@ -46,6 +49,13 @@ MenuView {
 
     signal loaded()
 
+    onNotationViewNavigationSectionChanged: function() {
+        menuNavPanel.section = root.notationViewNavigationSection
+    }
+    onNavigationOrderStartChanged: function() {
+        menuNavPanel.order = root.navigationOrderStart
+    }
```
Comment (cbjeukendrup) [inline src/framework/uicomponents/qml/Muse/UiComponents/internal/StyledMenu.qml:57]: This isn't very idiomatic QML: instead of setting the property imperatively to a fixed value, a binding (inside the `NavigationPanel` would be better:
```qml
    section: root.navigationSectionOverride ?? content.navigationSection
    order: root.navigationOrderStart
```

(which would remove the need for the addition of the `id: menuNavPanel` line)

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+StyledPopupView {
+    id: root
+    margins: 0
+
+    property NavigationSection notationViewNavigationSection: null
+    property int navigationOrderStart: 0
+    property int navigationOrderEnd: partialTieNavPanel.order
+
+    property QtObject model: partialTiePopupModel
```
Comment (cbjeukendrup) [inline src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:39]: It might be good to make this and similar properties in other popups `readonly` and `alias`.

File: src/notation/view/widgets/editstyle.ui [hunk 9]
```diff
@@ -8106,6 +8106,77 @@
...
+                   <property name="text">
+                    <string>Minimum hanging tie length:</string>
+                   </property>
+                   <property name="buddy">
+                    <cstring>minHangingTieLength</cstring>
+                   </property>
+                  </widget>
+                 </item>
+                 <item>
+                  <widget class="QDoubleSpinBox" name="minHangingTieLength">
```
Comment (cbjeukendrup) [inline src/notation/view/widgets/editstyle.ui:8129]: It would be good to fix the keyboard navigation order too, to prevent a situation like https://github.com/musescore/MuseScore/issues/25530; see that issue for info

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            id: partialTiePopupModel
+
+            onItemRectChanged: function(rect) {
+                root.elementRectChanged(rect)
+            }
+
+            onItemsChanged: function() {
+                tieMenuLoader.items = partialTiePopupModel.items
+                tieMenuLoader.show(Qt.point(0, 0))
+            }
```
Comment (cbjeukendrup) [inline src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:78]: This could also be done in a more idiomatic way: 
- the `tieMenuLoader.items = partialTiePopupModel.items` assignment is not necessary, because inside `tieMenuLoader` you create a "binding" by writing `items: partialTiePopupModel.items`, which means that `tieMenuLoader.items` is automatically updated when `partialTiePopupModel.items` is updated.
- as a consequence, for the `.show(…)` call, it is better to write 
    ```qml
                    onItemsChanged: function() {
                        tieMenuLoader.show(Qt.point(0, 0))
                    }
    ```
    inside `tieMenuLoader`.

File: src/notation/qml/MuseScore/NotationScene/internal/ElementPopupLoader.qml [hunk 11]
```diff
@@ -71,7 +72,13 @@ Item {
     function show(elementType, elementRect) {
         close()
 
-        var popup = loader.loadPopup(prv.componentByType(elementType), elementRect)
+        var component = prv.componentByType(elementType)
+
+        var popup = loader.loadPopup(component, elementRect)
+        if (!popup.model.canOpen) {
```
Comment (cbjeukendrup) [inline src/notation/qml/MuseScore/NotationScene/internal/ElementPopupLoader.qml:78]: It would be more ideal to make this decision earlier, in C++, to avoid needlessly creating and destroying popup windows.

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    property NavigationSection notationViewNavigationSection: null
+    property int navigationOrderStart: 0
+    property int navigationOrderEnd: partialTieNavPanel.order
+
+    property QtObject model: partialTiePopupModel
+
+    showArrow: false
+
+    onOpened: {
+        tieMenuLoader.show(Qt.point(0, 0))
```
Comment (cbjeukendrup) [inline src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:44]: It is a little bit of an ugly hack to have a StyledPopupView with nothing in it, and then launch a ContextMenuLoader from it. It would be nicer if we can avoid the StyledPopupView. That would require a bit of not-entirely-trivial refactoring though in, `ElementPopupLoader`. It might be better to save that for a next PR. Also, we should check if this hack does not cause graphical weirdnesses, like an extra border somewhere, or double drop shadow in a corner; if it does, that might increase the priority of the follow-up refactoring PR a bit.

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        const popupHeight = tieMenuLoader.item.height + root.margins * 2 + root.padding * 2
+        root.x = partialTiePopupModel.dialogPosition.x - root.parent.x
+        root.y = partialTiePopupModel.dialogPosition.y - root.parent.y - (opensUp ? popupHeight : 0)
+        root.setOpensUpward(opensUp)
+    }
+
+    contentWidth: content.width
+    contentHeight: content.childrenRect.height
+
+    ColumnLayout {
```
Comment (cbjeukendrup) [inline src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:65]: Perhaps this should not be a ColumnLayout but a simple `Item`, or perhaps even make the `ContextMenuLoader` itself a direct child of `root`, if that works (putting the model and NavigationPanel inside the ContextMenuLoader).

**Review by cbjeukendrup — COMMENTED** [review]
Here are some first (somewhat superficial) review comments. I haven't been able yet to take a deeper look at some parts of the code, so I might add some more comments later; if you prefer, you can of course wait until then before fixing these first comments.

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml (hunk 12 above)
Comment (miiizen) [reply src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:44]: Yeah, I agree it's messy and not ideal.  I'd like to address it outside of this PR though.

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml (hunk 8 above)
Comment (miiizen) [reply src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:39]: `notationViewNavigationSection` and `navigationOrderStart` are set in ElementPopupLoader.qml, so can't be readonly.

**Comment by miiizen** [comment]
@cbjeukendrup I've made the changes requested - I'd like to make any other changes in future PRs so we can get this merged asap!

File: src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml (hunk 8 above)
Comment (cbjeukendrup) [reply src/notation/qml/MuseScore/NotationScene/internal/PartialTiePopup.qml:39]: Sorry, I was unclear; I was referring to the `model` properties specifically here.

**Review by cbjeukendrup — COMMENTED** [review]
A few more comments; feel free to merge this PR and address them later. 

It would be great if you could provide an explanation of the structure of the PR: which information is stored how and where; how is the 'tree of ownership' (e.g. a Note owns a PartialTie), how are the crosslinks within that tree (e.g. a partial tie might have a reference to its counterpart(s)); etc.. 
Of course I should be able to extract that information from the code itself, but due to the size of the code, that is quite difficult in this case, both mentally and practically. So a small explanation to help me interpret what I'm seeing would increase efficiency a lot, to avoid constantly having to scroll up and down and use Ctrl+F all the time.

That said, I've not been able to look at every line of this PR; even if I'd do a local-only review, i.e. without trying to get a full understanding of the structure of the whole PR, it would take me some hours. So I've scanned it and commented on a few things that caught my eye. But again, feel free to address them in later PRs indeed.

**Comment by miiizen** [comment]
@cbjeukendrup hope this clears some things up.  This is just information on the relationship between elements - who owns what, and why certain pointers are being stored.
### PartialTie
`PartialTie` and `PartialTieSegment` are children of `Tie` and `TieSegment`.  They are added like normal ties to a `Note`, filling `m_tieFor` and `m_tieBack`. They behave in exactly the same way as regular ties, only without an end note.

### Tie
`Tie`s can act either as outgoing start ties, or incoming end ties.  
### tieJumpPoints()
This returns the `TieJumpPointList` belonging to the  `Tie`'s parent note.  This is only used when the tie is acting as a start tie
Screenshot 2025-01-07 at 16 03 01

### setJumpPoint() & jumpPoint()
This sets `m_jumpPoint` which is a `TieJumpPoint`.  This is only used when the tie is acting as an end tie. The tie needs to know about its `TieJumpPoint` to find its start tie

### TieJumpPoint
#### note()
This returns `m_note` which is a pointer to a `Note`. This is the note the end tie belongs to.
#### endTie()
This returns `tieBack()` of the note.
#### jumpPointList()
This returns a pointer to the `JumpPointList` that this object belongs to.  This information is needed so a jump point can find its start tie. Storing a pointer to a `Tie` would be more complex, as this `Tie` can change (eg. a `Tie` could be swapped for a `PartialTie` by unchecking "Tie to next note"). Each end point would then have to be updated.
This is needed for repitching and removing the start tie if all available endpoints have been removed (`Tie::updateStartTieOnRemoval`)

### TieJumpPointList
Each note has a `TieJumpPointList`.  This list needs to be managed by the `Note` as it lives longer than a `Tie` (eg. a `Tie` could be swapped for a `PartialTie` by unchecking "Tie to next note").
The `TieJumpPointList` stores a vector of `TieJumpPoint`s and a pointer to the start `Note`.  
#### m_jumpPoints
`TieJumpPoint`s are created in `Tie::updatePossibleJumpPoints` and added to the `TieJumpPointList` where they are managed. They are deleted in `~TieJumpPointLists`, or in `TieJumpPointList::clear`.  
#### startTie()
This returns the start tie.  This is the `tieFor` of the `Note` that this list belongs to. The note is stored in `m_note` and is only set on creation (as the `TieJumpPointList` is completely managed by the note)

### undoAddTieToScore(TieJumpPoint*) & undoRemoveTieFromScore(TieJumpPoint*)
These methods create the correct type of tie and add them to the score. 

---
So in short - ownership & management of `Tie`s and `PartialTie`s is unchanged (through the note). 
Ownership & management of `TieJumpPointList` is done by `Note`. 
Ownership and management of `TieJumpPoint` is done by `TieJumpPointList`
